File: java/src/org/doorways/t9/Numeric.java
package org.doorways.t9;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Numeric {
    static final String VERSION = "1.0.0";
    static final Map<String, List<String>> T9 = Map.of(
            "0", List.of("0"),
            "1", List.of("1"),
            "2", List.of("A", "B", "C"),
            "3", List.of("D", "E", "F"),
            "4", List.of("G", "H", "I"),
            "5", List.of("J", "K", "L"),
            "6", List.of("M", "N", "O"),
            "7", List.of("P", "Q", "R", "S"),
            "8", List.of("T", "U", "V"),
            "9", List.of("W", "X", "Y", "Z"));

    private static class Wheel {
        int click = 0;
        List<String> chars;

        Wheel(List<String> chars) {
            this.chars = chars;
        }
    }

    private static Queue<String> numeric(String digits) {
        List<List<String>> columns = new ArrayList<>(digits.length());
        int maximum = 1;

        //  Build the digit base map and character mapping
        //  based on the characters mapped from the original string.
        for (int i = 0; i < digits.length(); i++) {
            String focus = digits.substring(i, i + 1);
            List<String> chars = T9.getOrDefault(focus, List.of(focus));

            columns.add(chars);
            maximum *= chars.size();
        }

        Queue<String> result = new LinkedList<>();

        // Now just count.
        for (int i = 0; i < maximum; i++) {
            LinkedList<String> alpha = new LinkedList<>();
            int number = i;

            for (int col = columns.size() - 1; col >= 0; col--) {
                List<String> chars = columns.get(col);

                alpha.addFirst(chars.get(number % chars.size()));
                number /= chars.size();
            }

            result.add(String.join("", alpha));
        }

        return result;
    }

    public static void main(String[] args) {
        System.err.printf("### Starting Numeric.java %s\n", VERSION);

        try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim();

                Queue<String> result = numeric(line);

                System.out.printf("%-25s%7d result%s\n", line, result.size(), result.size() == 1 ? "" : "s");

                double count = 79.0 / (line.length() + 1.0);

                while (!result.isEmpty()) {
                    System.out.print(" ");

                    for (int i = 0; i < count; i++) {
                        if (result.isEmpty()) {
                            break;
                        }

                        System.out.printf(" %s", result.remove());
                    }

                    System.out.println();
                }
            }
        } catch (IOException e) {
            System.err.printf("*** Error reading from STDIN:  %s\n", e);
        }

        System.err.printf("### Finished Numeric.java %s\n", VERSION);
    }
}

File: go/numeric/numeric.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// Filters strings of digits into alphanumeric string via the T9 telephone keypad.
//
// Read strings of digits from STDIN, one line at a time.
// Convert each string of digits into a set of alphanumeric strings via the T9 telephone keypand
// using numeric algorithm and write the set of results to STDOUT.

var T9 = map[string][]string{
	"0": {"0"},
	"1": {"1"},
	"2": {"A", "B", "C"},
	"3": {"D", "E", "F"},
	"4": {"G", "H", "I"},
	"5": {"J", "K", "L"},
	"6": {"M", "N", "O"},
	"7": {"P", "Q", "R", "S"},
	"8": {"T", "U", "V"},
	"9": {"W", "X", "Y", "Z"},
}

const VERSION = "1.0.0"

//////////////////////////////////////////////////////////////////////////////
func numeric(digits string) []string {
	columns := make([][]string, len(digits))
	maximum := 1

	// Build the digit base map and character mapping
	// based on the characters mapped from the original string.
	for index, char := range digits {
		columns[index] = T9[string(char)]

		if len(columns[index]) < 1 {
			columns[index] = []string{string(char)}
		}

		maximum *= len(columns[index])
	}

	result := make([]string, 0)

	// Now just count.
	for i := 0; i < maximum; i++ {
		number := i
		temp := make([]string, 0)

		// From right to left pick character based on modulus by divisor
		// and divide number by divisor.
		for col := len(columns) - 1; col >= 0; col-- {
			chars := columns[col]
			divisor := len(chars)

			temp = append([]string{chars[number%divisor]}, temp...)
			number /= divisor
		}

		result = append(result, strings.Join(temp, ""))
	}

	return result
}

//////////////////////////////////////////////////////////////////////////////
func main() {
	_, _ = fmt.Fprintf(os.Stderr, "### Starting numeric.go %s\n", VERSION)

	scanner := bufio.NewScanner(os.Stdin)

	for scanner.Scan() {
		digits := strings.TrimSpace(scanner.Text())
		result := numeric(digits)
		rCount := len(result)
		plural := "s"

		if rCount == 1 {
			plural = ""
		}

		fmt.Printf("%-25s%7d result%s\n", digits, rCount, plural)

		count := 79 / float64(len(digits)+1)

		for len(result) > 0 {
			fmt.Print(" ")

			for i := 0; float64(i) < count; i++ {
				if len(result) < 1 {
					break
				}

				fmt.Printf(" %s", result[0])
				result = result[1:]
			}

			fmt.Println()
		}
	}

	_, _ = fmt.Fprintf(os.Stderr, "### Finished numeric.go %s\n", VERSION)
}
